lean/main/metam.lean

```diff
@@ -796,189 +796,300 @@ factors:
+-- fun {α : Type u_1} (xs ys : List.{u_1} α) => @List.append.{u_1} α (@List.append.{u_1} α xs ys) xs
 
-/-! One of the first ways to construct expressions is by function application, that is, to form the expression for `f a` given expressions for `f` and `a`. One uses various methods provided by Lean 4 to do this. It is generally not a good idea to try to directly define the expression, as there is associated metadata that needs to be included which is difficult to construct manually.
-
-The simple methods for constructing expressions by function application are `mkApp` and `mkAppN`. These take care of the metadata, but do not attempt to unify, for example. The `mkAppN` function takes a function and an array of arguments. The name ending with `N` here (and elsewhere) is to indicate that we (in general) have multiple (`N`) arguments. On the other hand `mkApp` applies a function to a single argument.
+/-!
+The `.{u_1}` suffixes are universe levels, which must be given for every
+polymorphic constant. And of course the type `α` is passed around everywhere.
 
-The following example illustrates constructing examples using `mkAppN`. This
-would yield an even longer expression than the previous example, but we again
-use `reduce` to simplify it:
+Exactly the same problem occurs during metaprogramming when we construct
+expressions. A hand-made expression representing the right-hand side of the
+above definition looks like this:
 -/
 
-def sumExprM (n m : Nat) : MetaM Expr := do
-  reduce $ mkAppN (mkConst ``Nat.add) #[natExpr n, natExpr m]
+def appendAppendRHSExpr₁ (u : Level) (α xs ys : Expr) : Expr :=
+  mkAppN (mkConst ``List.append [u])
+    #[α, mkAppN (mkConst ``List.append [u]) #[α, xs, ys], xs]
 
-#eval sumExprM 2 3 --Lean.Expr.lit (Lean.Literal.natVal 5) (Expr.mkData 1441793 (bi := Lean.BinderInfo.default))
+/-!
+Having to specify the implicit arguments and universe levels is annoying and
+error-prone. So `MetaM` provides a helper function which allows us to omit
+implicit information: `Lean.Meta.mkAppM` of type
 
-/-! Note that the _function_ to be applied is a constant, and so has an expression of the form `mkConst name`. The name was given preceded by double-backticks. This means that Lean resolves this to a global name, and gives an error if it cannot be resolved. A name preceded by a single backtick is a literal name.
--/
+```lean
+mkAppM : Name → Array Expr → MetaM Expr
+```
 
-/-! Besides function application, we would like to construct expressions using λ-expressions. 
-We next illustrate how to do this by constructing a λ-expression for the function `double : Nat → Nat` given by `double n = n + n`. 
+With this, the example becomes much shorter:
+-/
 
-To construct such an expression, we _introduce_ a free variable `n`, define an expression in terms of this variable, and finally construct the λ-expression. 
+def appendAppendRHSExpr₂ (xs ys : Expr) : MetaM Expr := do
+  mkAppM ``List.append #[← mkAppM ``List.append #[xs, ys], xs]
```

Comment:
This is a good opportunity to be bolder about the difference between `mkAppN` and `mkAppM`. Patrick said that it wasn't completely clear for him. Maybe even mention "N" vs "M" to help the reader, as that detail can be hard to grasp in the middle of such a complex topic already